Libraries/LibCrypto/Cipher/Mode/Mode.h

```diff
@@ -0,0 +1,105 @@
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <AK/ByteBuffer.h>
+#include <LibCrypto/Cipher/Cipher.h>
+
+namespace Crypto {
+namespace Cipher {
+
+template<typename T>
+class Mode {
+public:
+    virtual ~Mode() {}
+
+    // FIXME: Somehow communicate that encrypt returns the last initialization vector (if the mode supports it)
+    virtual Optional<ByteBuffer> encrypt(const ByteBuffer& in, ByteBuffer& out, Optional<ByteBuffer> ivec = {}) = 0;
+    virtual void decrypt(const ByteBuffer& in, ByteBuffer& out, Optional<ByteBuffer> ivec = {}) = 0;
+
+    virtual size_t IV_length() const = 0;
+
+    const T& cipher() const { return m_cipher; }
+
+    ByteBuffer create_aligned_buffer(size_t input_size) const
+    {
+        size_t remainder = (input_size + T::block_size()) % T::block_size();
+        if (remainder == 0)
+            return ByteBuffer::create_uninitialized(input_size);
+        else
+            return ByteBuffer::create_uninitialized(input_size + T::block_size() - remainder);
+    }
+
+    virtual String class_name() const = 0;
+    T& cipher() { return m_cipher; }
+
+protected:
+    virtual void prune_padding(ByteBuffer& data)
+    {
+        auto size = data.size();
+        switch (m_cipher.padding_mode()) {
+        case PaddingMode::CMS: {
+            auto maybe_padding_length = data[size - 1];
+            if (maybe_padding_length >= T::block_size() - 1) {
+                // cannot be padding (the entire block cannot be padding)
+                return;
+            }
+            for (auto i = maybe_padding_length; i > 0; --i) {
+                if (data[size - i - 1] != maybe_padding_length) {
```

Comment:
```suggestion
            for (auto i = size - maybe_padding_length; i < size; ++i) {
                if (data[i] != maybe_padding_length) {
```

The current code has an off-by-one error due to the same issue as with the encryption (CMS padding is actually "data_size - length" without any -1).

The simpler fix is I think to stop the iteration to `i > 1`.

Styling-wise, I propose reworking the loop like this. It's functionally equivalent to the existing one, but I find it more intuitive to think about iterating from "somewhere in the middle of the block" to "the end of the block", rather than how it's written right now (backwards over the length of the padding but starting from the end of the block).

Reply:
I forgot to mention this in the review, but the `data.trim()` line right below this should also be changed to match the value : `data.trim(size - maybe_padding_length - 1);` => `data.trim(size - maybe_padding_length);`